Data/Primitive/Types.hs

```diff
@@ -390,6 +392,36 @@ deriving instance Prim CTimer
 #endif
 deriving instance Prim Fd
 
+-- | @since 0.7.1.0
+instance Prim WordPtr where
+  sizeOf# _ = sizeOf# (undefined :: Ptr ()) 
+  alignment# _ = alignment# (undefined :: Ptr ()) 
+  indexByteArray# a i = ptrToWordPtr (indexByteArray# a i)
+  readByteArray# a i s0 = case readByteArray# a i s0 of
+    (# s1, p #) -> (# s1, ptrToWordPtr p #)
+  writeByteArray# a i wp = writeByteArray# a i (wordPtrToPtr wp)
+  setByteArray# a i n wp = setByteArray# a i n (wordPtrToPtr wp)
+  indexOffAddr# a i = ptrToWordPtr (indexOffAddr# a i)
+  readOffAddr# a i s0 = case readOffAddr# a i s0 of
+    (# s1, p #) -> (# s1, ptrToWordPtr p #)
+  writeOffAddr# a i wp = writeOffAddr# a i (wordPtrToPtr wp)
+  setOffAddr# a i n wp = setOffAddr# a i n (wordPtrToPtr wp)
```

Comment:
Given the large amount of code repetition here, perhaps we should factor this amount into a CPP macro? The existing `derivePrim` is _almost_ what we want, although there are enough differences between the code you're using here and `derivePrim` where I think it would make sense for this to exist as its own macro.

Reply:
With some clever internal use of pattern synonyms, we could actually reuse `derivePrim` as is. But I think it's better to just have a different macro as you suggest. I'll look into this.